src/ray/stats/metric.cc

```diff
@@ -114,25 +112,20 @@ void Metric::Record(double value, const TagsType &tags) {
   if (StatsConfig::instance().IsStatsDisabled()) {
     return;
   }
-  // NOTE(lingxuan.zlx): Double check for recording performance while
-  // processing in multithread and avoid race since metrics may invoke
-  // record in different threads or code pathes.
-  if (measure_ == nullptr) {
-    absl::MutexLock lock(&registration_mutex_);
-    if (measure_ == nullptr) {
-      // Measure could be registered before, so we try to get it first.
-      MeasureDouble registered_measure =
-          opencensus::stats::MeasureRegistry::GetMeasureDoubleByName(name_);
-
-      if (registered_measure.IsValid()) {
-        measure_.reset(new MeasureDouble(registered_measure));
-      } else {
-        measure_.reset(
-            new MeasureDouble(MeasureDouble::Register(name_, description_, unit_)));
-      }
-      RegisterView();
+
+  std::call_once(measure_init_flag_, [this]() {
+    // Measure could be registered before, so we try to get it first.
+    MeasureDouble registered_measure =
+        opencensus::stats::MeasureRegistry::GetMeasureDoubleByName(name_);
```

Comment:
Hm, do you know why this part was needed in the first place and if it's still necessary? Perhaps because there can be multiple metric objects created with the same name?

Reply:
`Record` is used to export metrics, my guess is lazy initialization for performance.

Reply:
I'm wondering why we need to handle the case where the measure is already registered if we're guarding it with `std::call_once`

Reply:
I don't understand what you mean, could you please elaborate?

My guess is, people don't want to eager initialize measure and register view because in certain cases such initialization could be skippped.

Reply:
That part makes sense.

But why do we need to guard against the case where the metric has already been registered? See the comment above this line and logic handling it below

Reply:
It's used in unit test, https://github.com/ray-project/ray/blob/51c35af010357fadd530ca801600dc1093852270/src/ray/stats/stats_test.cc#L133-L134

where we clean up shutdown all opencensus resource later re-init and re-register again.

Reply:
I don't think it's something would happen in real production.